atomix/cluster/src/test/java/io/atomix/raft/impl/PriorityElectionTimerTest.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Copyright © 2020 camunda services GmbH (info@camunda.com)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.atomix.raft.impl;
+
+import io.atomix.utils.concurrent.SingleThreadContext;
+import java.time.Duration;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+import org.awaitility.Awaitility;
+import org.junit.After;
+import org.junit.Test;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class PriorityElectionTimerTest {
+
+  private final Logger log = LoggerFactory.getLogger(PriorityElectionTimerTest.class);
+  private final SingleThreadContext threadContext = new SingleThreadContext("priorityElectionTest");
+
+  @After
+  public void after() {
+    threadContext.close();
+  }
+
+  @Test
+  public void shouldLowerPriorityNodeEventuallyStartsAnElection() {
+    // given
+    final AtomicInteger triggerCount = new AtomicInteger();
+
+    final PriorityElectionTimer timer =
+        new PriorityElectionTimer(
+            Duration.ofMillis(100), threadContext, triggerCount::getAndIncrement, log, 4, 1);
+
+    // when
+    timer.reset();
+
+    // then
+    Awaitility.await().until(() -> triggerCount.get() >= 1);
+  }
+
+  @Test
+  public void shouldHighPriorityNodeStartElectionFirst() {
```

Comment:
Took me  a while to get this test :sweat_smile:

Reply:
:see_no_evil: Any suggestions to improve it?

Reply:
Unfortunately, when I was writing the first comment I had not really an idea to improve this. I just wanted to note it.

Don't get me wrong in general if you get the test it is quite elegant :) I also thought about having latches, where you can verify the order of calls, but it makes even more complex.

During writing this comment I had another idea :D I think here would be a good place to use mocks :D You can use two mocks and then the InOrder verification. What do you think?

Reply:
Thanks. I will try out your idea with mocks. I will merge this PR now. But will create a followup PR to improve the test.